### PR TITLE
Testing MiMC solidity implementation

### PR DESCRIPTION
**Pull request by fredlacs**
Testing of MiMC solidity implementation against js library.
TODO: parse it with ying tong's MiMCMerkle.js after merge and check the multihash.

### CODE REVIEW

File: test/mimc.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        
+        const solF = await mimc.methods.MiMCpe7(c,d).call();
+        const jsF = mimcjs.hash(c,d,91);
+        assert.equal(solF.toString(), jsF.toString());
+        
+        const solG = await mimc.methods.MiMCpe7(solE,solF).call();
+        const jsG = mimcjs.hash(jsE,jsF,91);
+        assert.equal(solG.toString(), jsG.toString());
+
+        // TODO: parse it with ying tong's MiMCMerkle.js after merge
```
Comment (onyb) [inline test/mimc.js]: @therealyingtong Do you think we'll be able to use the `MiMCMerkle` contract from ethsnarks? We could simply extend the contract and provide our own version of `HashImpl` to make it work with `mimc7.js` from circomlib.

See: https://github.com/HarryR/ethsnarks/blob/master/contracts/MerkleTree.sol#L58

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,5 @@
 node_modules/
 .secret
 proving_key.json
-verification_key.json
\ No newline at end of file
+verification_key.json
+build/
```
Comment (onyb) [inline .gitignore]: How come `build` is gitignored, but I'm still seeing the JSON files inside `build` in this pull request?

File: test/mimc.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            E    F
+          A  B  C  D
+
+        E = hash(A, B)
+        F = hash(C, D)
+        G = hash(E, F)
+
+        multihash(a,b,c,d) != G
+        multihash(a,b,c,d) != hash(hash(a, b), hash(c, d))
+        multihash(a,b,c,d) == hash(hash(hash(hash(IV, a), b), c), d)
```
Comment (onyb) [inline test/mimc.js]:
```
              G          <-- merkle root
            E    F
          A  B  C  D     <-- merkle tree leaves
```
Consider the above Merkle tree, where `E = hash(A, B)`, `F = hash(C, D)`, and `G = hash(E, F)`.

We could have a function called `multiHashTree`, such that `multiHashTree(A, B, C, D) == G`, noting that `G` is the hash of the Merkle tree root. Is this something we might need?

File: test/mimc.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        assert.equal(solF.toString(), jsF.toString());
+        
+        // TODO: parse it with ying tong's MiMCMerkle.js after merge and assert values
+        // g = hash(hash(a, b), hash(c, d))
+        const solG = await mimc.methods.MiMCpe7(solE, solF).call();
+        const jsG = mimcjs.hash(jsE, jsF);
+        assert.equal(solG.toString(), jsG.toString());
+
+        // multihash = hash(hash(hash(hash(IV, a), b), c), d)
+        const multiHashDecomposition = await solidityHash(await solidityHash(await solidityHash(await solidityHash(mimcjs.getIV(), a), b), c), d);
```
Comment (onyb) [inline test/mimc.js]: TODO: Check if there is a way to get the value of `IV` from the MiMC contract, and use it here instead.

**Comment by barryWhiteHat** [comment]
@fredlacs @onyb It feels like this is still being reviewed. If you ping me when this is reviewd i can merge.

File: .gitignore (hunk 2 above)
Comment (fredlacs) [reply .gitignore]: It was pushed previously to avoid us having to compile the circuits over and over, I believe.

File: test/mimc.js (hunk 1 above)
Comment (fredlacs) [reply test/mimc.js]: If we do end up using ethsnarks MiMC implementation, could be nice to also use their MiMCMerkle.

File: test/mimc.js (hunk 3 above)
Comment (fredlacs) [reply test/mimc.js]: Could be useful, and we can easily implement it once the need comes!

**Comment by therealyingtong** [comment]
thanks for your work on this @fredlacs @onyb, it helped greatly in getting a working MiMCMerkle.sol on chain :-) closing this to keep the open PRs updated

**Comment by fredlacs** [comment]
> thanks for your work on this @fredlacs @onyb, it helped greatly in getting a working MiMCMerkle.sol on chain :-) closing this to keep the open PRs updated

No problem, glad to have helped! :)
Btw, great progress has been going on, gj!

**Pull request closed without merging**